src/compiler/crystal/semantic/exhaustiveness_checker.cr

```diff
@@ -0,0 +1,162 @@
+struct Crystal::ExhaustivenessChecker
+  def initialize(@program : Program)
+  end
+
+  def check(node : Case)
+    cond = node.cond
+
+    # No condition means it's just like a series of if/else
+    return unless cond
```

Comment:
You are right! For the cond-less `case` I think we should also make it exhaustive. That way `case` always means exhaustive. I'll change it 👍

Reply:
To confirm: this is only talking about a few edge cases, right?

All cases warn if they're non-exhaustive *unless* they have no condition or they use tuple pattern matching after this PR?

Reply:
I was probably not clear. I want all `case cond` to behave the same. So, if the exhaustivity can be proved, the `else` is not needed. Otherwise (if the exhaustivity could not be proved either because of a missing `when` or because the checked can't compute the proof for that type) it should warn.

This means that if later on a proof for that condition type is added the programmer or a tool will need to highlight the unneeded `else`; but I prefer that that having different semantics for `case cond`.

The cond-less `case` does notnecessarilyy fit in this definition, since there is no cond type. So for consistency we can warn; or for convenience as with else-less `if`, we can have an implicit `else nil`. This is the part I am hessitant.

Reply:
Yes. Though I'll make sure to make **all** case exhaustive. I'll implement it for tuple literals too.

Reply:
Hm, the comment made it to a different thread. Disregard it.

Reply:
Let me see if I understand. If there's a `case` that covers all cases through `when`, but it still has an `else`, the compiler should give a warning/error saying you need to remove the `else`?

I can do that. Do you want me to do it in this PR or in a separate one? (or not at all?)

> The cond-less case does notnecessarilyy fit in this definition, since there is no cond type. So for consistency we can warn; or for convenience as with else-less if, we can have an implicit else nil. This is the part I am hessitant.

For consistency, I think it's better to make all `case` be exhaustive. Otherwise it can be a bit confusing.

Maybe later on we can remove the cond-less case if it's just another way to write `if / elsif`. We copied it from Ruby, but in my experience not many know about this special notation, so the simpler we make the language, the better.

Reply:
> If there's a case that covers all cases through when, but it still has an else ..

I don't mean to do something now. But what will happen to the code if additional exhaustiveness checks would've been added later. No need to do anything in this PR (or future PRs in the compiler IMO).

Reply:
> a tool will need to highlight the unneeded `else`

Consider this example:

```crystal
def foo(x)
  case x
  when Int32
    # ...
  when String
    # ...
  else
    # ...
  end
end
```

Here I coded `foo` with `Int32`, `String` and other types in mind. If I call it like this:

```crystal
foo(1 || "hello" || true)
```

then it works fine.

However, if I call it like this:

```crystal
foo(1)
```

then, if we make the compiler say "else is not needed if you already covered all types" then the compiler will complain, but that's not correct. It could even complain because `1` will never match `String` but that's incorrect too. But only because of the way we instantiate methods.

So I think we should only make `case` exhaustive and not require an `else` in that case, but we can't complain about extra types that will never match, or an `else` that will never be reached (because we really don't know if it really can't be reached).

Reply:
I agree, we should not complain about extra `else`. I would handle dead code elimination outside the compiler.

My example was in an scenario where a new check is added in the compiler, after this is merged, hence the code will have an extra `else` since all the cases are exhaustive. But I am satisfied with the current state of that story.